### PR TITLE
Implement full EVM-C interface

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: libevmjit/Cache.cpp [hunk 1]
```diff
@@ -128,13 +129,12 @@ std::unique_ptr<llvm::Module> Cache::getObject(std::string const& id)
 
 	if (g_lastObject)  // if object found create fake module
 	{
-		DLOG(cache) << id << ": found\n";
-		auto&& context = llvm::getGlobalContext();
-		auto module = llvm::make_unique<llvm::Module>(id, context);
-		auto mainFuncType = llvm::FunctionType::get(llvm::Type::getVoidTy(context), {}, false);
+        DLOG(cache) << id << ": found\n";
```
Comment (chriseth) [inline libevmjit/Cache.cpp]: tabs-spaces

File: libevmjit/Compiler.cpp [hunk 2]
```diff
@@ -783,8 +783,8 @@ void Compiler::compileBasicBlock(BasicBlock& _basicBlock, RuntimeManager& _runti
 			auto callGas = stack.pop();
 			auto address = stack.pop();
 			auto value = (kind == EVM_DELEGATECALL) ?
-							 llvm::UndefValue::get(Type::Word) :
-							 stack.pop();
+			             llvm::UndefValue::get(Type::Word) :
```
Comment (chriseth) [inline libevmjit/Compiler.cpp]: tabs/spaces

File: libevmjit/JIT.cpp [hunk 3]
```diff
@@ -31,18 +30,18 @@ namespace
 {
 using ExecFunc = ReturnCode(*)(ExecutionContext*);
 
-template <size_t _size>
-std::string toHex(std::array<byte, _size> const& _data)
+/// Combine code hash and compatibility mode into a printable code identifier.
+std::string makeCodeId(evm_hash256 codeHash, evm_mode mode)
```
Comment (chriseth) [inline libevmjit/JIT.cpp]: Pass by const ref?

File: libevmjit/JIT.cpp [hunk 4]
```diff
@@ -31,18 +30,18 @@ namespace
...
 	std::string str;
-	str.resize(_size * 2);
-	auto outIt = str.rbegin(); // reverse for BE
-	for (auto b: _data)
+	str.reserve(sizeof(codeHash) * 2 + 1);
+	for (auto b: codeHash.bytes)
 	{
-		*(outIt++) = hexChars[b & 0xf];
-		*(outIt++) = hexChars[b >> 4];
+		str.push_back(hexChars[b & 0xf]);
```
Comment (chriseth) [inline libevmjit/JIT.cpp]: Cast b to make sure it is unsigned?

File: libevmjit/JIT.cpp [hunk 5]
```diff
@@ -207,122 +221,137 @@ ExecFunc JITImpl::compile(byte const* _code, uint64_t _codeSize, std::string con
...
+	return reinterpret_cast<evm_instance*>(&jit);
+}
+
+EXPORT void evm_destroy(evm_instance* instance)
+{
+	assert(instance == static_cast<void*>(&JITImpl::instance()));
+}
+
+EXPORT evm_result evm_execute(evm_instance* instance, evm_env* env,
+                                  evm_mode mode, evm_hash256 code_hash,
```
Comment (chriseth) [inline libevmjit/JIT.cpp]: Indentation looks weird. If you always use exactly a single tab, it never looks weird ;-)

File: libevmjit/JIT.cpp [hunk 6]
```diff
@@ -207,122 +221,137 @@ ExecFunc JITImpl::compile(byte const* _code, uint64_t _codeSize, std::string con
...
 {
-	int64_t scheduleId = id();
-	return
-		toHex(*(std::array<byte, 32>*)&_codeHash) +
-		"-" +
-		toHex(*(std::array<byte, 8>*)&scheduleId);
+	auto& jit = *reinterpret_cast<JITImpl*>(instance);
+	auto codeIdentifier = makeCodeId(code_hash, mode);
+	auto execFunc = jit.compile(mode, code, code_size, codeIdentifier);
+	if (execFunc) // FIXME: What with error?
```
Comment (chriseth) [inline libevmjit/JIT.cpp]: In my personal opinion, using an auto-variable in a boolean context is very dangerous. The type checking will often not complain about a change of the return type of `jit.compile` and thus a required matching change at this point might be forgotten.

Comment (chfast) [inline libevmjit/JIT.cpp]: I will keep that in mind. However, in this case the `execFunc` is a pointer to function.

**Comment by chfast** [comment]
Sorry about wasting our time discussing spaces, but I use 5 different instances of code editors and they try to be smart about the formatting. I also have to switch to LLVM quite often. This is the reason I push to have code format rules that are at least verifiable by any tool.

File: libevmjit/JIT.cpp (hunk 3 above)
Comment (chfast) [inline libevmjit/JIT.cpp]: I will benchmark it later.

Comment (chfast) [inline libevmjit/JIT.cpp]: The performance is more/less the same. I prefer passing by value, but looks like in other code ref is used. What's your opinion?

File: libevmjit/JIT.cpp (hunk 6 above)
Comment (chfast) [inline libevmjit/JIT.cpp]: Is the explanation enough?

Comment (chriseth) [inline libevmjit/JIT.cpp]: The problem is that "this case" can change any time. You might change it to a function reference or whatever.

**Pull request merged**